### PR TITLE
[SPARK-47408][SQL] Fix mathExpressions that use StringType

### PR DESCRIPTION
**Pull request by mihailomilosevic2001**
### What changes were proposed in this pull request?
Support more functions that use strings with collations.

### Why are the changes needed?
Hex, Unhex, Conv are widely used and need to be enabled wih collations

### Does this PR introduce _any_ user-facing change?
Yes, enabled more functions.

### How was this patch tested?
With new tests in `CollationSQLExpressionsSuite.scala`.

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

**Review by uros-db — APPROVED** [review]
lgtm

**Comment by HyukjinKwon** [comment]
Can you create a JIRA, and link it to the PR title please?

File: sql/core/src/test/scala/org/apache/spark/sql/CollationSQLExpressionsSuite.scala [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+    val testCases = Seq(
+      ConvTestCase("100", "2", "10", "UTF8_BINARY", "4"),
+      ConvTestCase("100", "2", "10", "UTF8_BINARY_LCASE", "4"),
+      ConvTestCase("100", "2", "10", "UNICODE", "4"),
+      ConvTestCase("100", "2", "10", "UNICODE_CI", "4")
+    )
+    testCases.foreach(t => {
+      val query =
+        s"""
+           |select conv(collate('${t.num}', '${t.collationName}'), ${t.from_base}, ${t.to_base})
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/CollationSQLExpressionsSuite.scala:49]: why do we add end-to-end tests but not unit tests?

Comment (uros-db) [reply sql/core/src/test/scala/org/apache/spark/sql/CollationSQLExpressionsSuite.scala:49]: I think these expressions are only pass-through (changes only in `inputTypes`/`dataType`) in mathExpressions.scala, so these tests should be enough

**Review by cloud-fan — APPROVED** [review]

**Review by dongjoon-hyun — APPROVED** [review]

**Comment by dongjoon-hyun** [comment]
Merged to master.

**Pull request closed without merging**